### PR TITLE
rustdoc: add header map to the table of contents

### PR DESCRIPTION
**Pull request by notriddle**
## Summary

Add header sections to the sidebar TOC.

### Preview

image

* http://notriddle.com/rustdoc-html-demo-9/toc/rust/std/index.html
* http://notriddle.com/rustdoc-html-demo-9/toc/rust-derive-builder/derive_builder/index.html

## Motivation

Some pages are very wordy, like these.

| crate | word count |
|--|--|
| [std::option](https://doc.rust-lang.org/stable/std/option/index.html) | 2,138
| [derive_builder](https://docs.rs/derive_builder/0.13.0/derive_builder/index.html) | 2,403
| [tracing](https://docs.rs/tracing/0.1.40/tracing/index.html) | 3,912
| [regex](https://docs.rs/regex/1.10.3/regex/index.html) | 8,412

This kind of very long document is more navigable with a table of contents, like Wikipedia's or the one [GitHub recently added](https://github.blog/changelog/2021-04-13-table-of-contents-support-in-markdown-files/) for READMEs.

In fact, the use case is so compelling, that it's been requested multiple times and implemented in an extension:

* https://github.com/rust-lang/rust/issues/80858
* https://github.com/rust-lang/rust/issues/28056
* https://github.com/rust-lang/rust/issues/14475
* https://rust.extension.sh/#show-table-of-content

(Some of these issues ask for more than this, so don’t close them.)

It's also been implemented by hand in some crates, because the author really thought it was needed. Protip: for a more exhaustive list, run [`site:docs.rs table of contents`](https://duckduckgo.com/?t=ffab&q=site%3Adocs.rs+table+of+contents&ia=web), though some of them are false positives.

* https://docs.rs/figment/0.10.14/figment/index.html#table-of-contents
* https://docs.rs/csv/1.3.0/csv/tutorial/index.html#table-of-contents
* https://docs.rs/axum/0.7.4/axum/response/index.html#table-of-contents
* https://docs.rs/regex-automata/0.4.5/regex_automata/index.html#table-of-contents

Unfortunately for these hand-built ToCs, because they're just part of the docs, there's no consistent way to turn them off if the reader doesn't want them. It's also more complicated to ensure they stay in sync with the docs they're supposed to describe, and they don't stay with you when you scroll like Wikipedia's [does now](https://uxdesign.cc/design-notes-on-the-2023-wikipedia-redesign-d6573b9af28d).

## Guide-level explanation

When writing docs for a top-level item, the first and second level of headers will be shown in an outline in the sidebar. In this context, "top level" means "not associated".

This means, if you're writing very long guides or explanations, and you want it to have a table of contents in the sidebar for its headings, the ideal place to attach it is usually the *module* or *crate*, because this page has fewer other things on it (and is the ideal place to describe "cross-cutting concerns" for its child items).

If you're reading documentation, and want to get rid of the table of contents, open the image Settings panel and checkmark "Hide table of contents."

## Reference-level explanation

Top-level items have an outline generated. This works for potentially-malformed header trees by pairing a header with the nearest header with a higher level. For example:

```markdown
## A
# B
# C
## D
## E
```

A, B, and C are all siblings, and D and E are children of C.

Rustdoc only presents two layers of tree, but it tracks up to the full depth of 6 while preparing it.

That means that these two doc comment both generate the same outline:

```rust
/// # First
/// ## Second
struct One;
/// ## First
/// ### Second
struct Two;
```

## Drawbacks

The biggest drawback is adding more stuff to the sidebar.

My crawl through docs.rs shows this to, surprisingly, be less of a problem than I thought. The manually-built tables of contents, and the pages with dozens of headers, usually seem to be modules or crates, not types (where extreme scrolling would become a problem, since they already have methods to deal with).

The best example of a type with many headers is [vec::Vec](https://doc.rust-lang.org/1.75.0/std/vec/struct.Vec.html), which still only has five headers, not dozens like [axum::extract](https://docs.rs/axum/0.7.4/axum/extract/index.html).

### Why in the existing sidebar?

The method links and the top-doc header links have more in common with each other than either of them do with the "In [parent module]" links, and should go together.

### Why limited to two levels?

The sidebar is pretty narrow, and I don't want too much space used by indentation. Making the sidebar wider, while it has some upsides, also takes up more space on middling-sized screens or tiled WMs.

### Why not line wrap?

That behaves strangely when resizing.

### Doc generators that have TOC for headers

https://hexdocs.pm/phoenix/Phoenix.Controller.html is very close, in the sense that it also has header sections directly alongside functions and types.

Another example, referenced as part of the [early sidebar discussion](https://github.com/rust-lang/rust/issues/37856) that added methods, Ruby will show a table of contents in the sidebar (for example, on the [ARGF](https://docs.ruby-lang.org/en/master/ARGF.html) class). According to their changelog, [they added it in 2013](https://github.com/ruby/rdoc/blob/06137bde8ccc48cd502bc28178bcd8f2dfe37624/History.rdoc#400--2013-02-24-).

Haskell seems to mix text and functions even more freely than Elixir. For example, this [Naming conventions](https://hackage.haskell.org/package/base-4.19.0.0/docs/Control-Monad.html#g:3) is plain text, and is immediately followed by functions. And the [Pandoc top level](https://hackage.haskell.org/package/pandoc-3.1.11.1/docs/Text-Pandoc.html) has items split up by function, rather than by kind. Their TOC matches exactly with the contents of the page.

### Doc generators that don't have header TOC, but still have headers

Elm, interestingly enough, seems to have the same setup that Rust used to have: sibling navigation between modules, and no index within a single page. [They keep Haskell's habit of named sections with machine-generated type signatures](https://package.elm-lang.org/packages/elm/browser/latest/Browser-Dom), though.

[PHP](https://www.php.net/manual/en/book.datetime.php), like elm, also has a right-hand sidebar with sibling navigation. However, PHP has a single page for a single method, unlike Rust's page for an entire "class." So even though these pages have headers, it's never more than ten at most. And when they have guides, those guides are also multi-page.

## Unresolved questions

* Writing recommendations for anyone who wants to take advantage of this.
* Right now, it does not line wrap. That might be a bad idea: a lot of these are getting truncated.
* Split sidebars, which I [tried implementing](https://rust-lang.zulipchat.com/#narrow/stream/266220-t-rustdoc/topic/Table.20of.20contents), are not required. The TOC can be turned off, if it's really a problem. Implemented in https://github.com/rust-lang/rust/pull/120818, but needs more, separate, discussion.

## Future possibilities

I would like to do a better job of distinguishing global navigation from local navigation. Rustdoc has a pretty reasonable information architecture, if only we did a better job of communicating it.

This PR aims, mostly, to help doc authors help their users by writing docs that can be more effectively skimmed. But it doesn't do anything to make it easier to tell the TOC and the Module Nav apart.

### CODE REVIEW

**Comment by the8472** [comment]
> In fact, the use case is so compelling, that it's been requested multiple times and implemented in an extension:

At least some of the requests (e.g. #14475, #40273) are about method signature summary tables, not the headings for the prose.

**Comment by the8472** [comment]
Regarding the issues. I want to avoid issues getting closed when this isn't the thing some of them are asking for. And also make it clear that what is being asked for is more complicated.

> Is it a blocking concern for this PR? 

No.

**Comment by GuillaumeGomez** [comment]
My only issue with the current version is it's actually not obvious what I'm reading at the top of the sidebar (also I'm a bit sad because I really liked the version with a new sidebar).

It's also a bit strange to have the table of contents followed directly by section links like "Structs" as there is no real visual marker to indicate it's not the table of content anymore.

**Comment by GuillaumeGomez** [comment]
> Would an explicit label for the "Sections" be a good idea, like how Elixir does it?

I think I prefer the non-wikipedia approach, but both look much better than the current situation!

> That's still the table of contents. They're programmatically generated, instead of coming from the Markdown text, but they're still part of the current page.

Then maybe add a bit more of spacing to make it more obvious it's not the same thing?

> I agree. https://github.com/rust-lang/rust/pull/120818 does that.

:+1:

**Comment by GuillaumeGomez** [comment]
My issue here is that it takes **me** (not sure about others, so might be irrelevant) much more time to find the methods/items list when I need to go through the TOC first. There is no visual marker to help there too.

**Comment by Folyd** [comment]
Good to know librustdoc support TOC in the doc page. Just a suggestion, why not put the TOC to the right? Rust Search Extension is an example :)
https://rust.extension.sh/#show-table-of-content

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Sorry it took me so long to come back to this PR. This is a nice improvement! Please start the FCP once merge conflicts are fixed.

**Comment by GuillaumeGomez** [comment]
I see in your demo that you have:

image

Is it coming from a change in this PR?

**Comment by GuillaumeGomez** [comment]
Ah nevermind, it was a bug on current nightly. Fixed in https://github.com/rust-lang/rust/pull/127963. Can you regen once this PR is merged please? Just want to do one last tour to ensure I didn't miss anything before approving.

File: src/librustdoc/html/templates/sidebar.html [hunk 1]
```diff
@@ -11,24 +6,44 @@ <h2 class="location"> {# #}
...
+                <a href="#">{{title_prefix}}{{title}}</a> {# #}
+            </h2>
+            {% endif %}
             {% for block in blocks %}
                 {% if block.should_render() %}
                     {% if !block.heading.name.is_empty() %}
-                        <h3><a href="#{{block.heading.href|safe}}">{{block.heading.name}}</a></h3>
+                        <h3> {# #}
+                            <a href="#{{block.heading.href|safe}}">{{block.heading.name}}</a> {# #}
+                        </h3> {# #}
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/templates/sidebar.html:20]: Can be:

```suggestion
                        </h3>
```

File: src/librustdoc/html/templates/sidebar.html [hunk 2]
```diff
@@ -11,24 +6,44 @@ <h2 class="location"> {# #}
...
                             {% for link in block.links %}
-                                <li><a href="#{{link.href|safe}}">{{link.name}}</a></li>
+                                <li> {# #}
+                                    <a href="#{{link.href|safe}}">{{link.name}}</a> {# #}
+                                    {% if !link.children.is_empty() %}
+                                        <ul> {# #}
+                                            {% for child in link.children %}
+                                                <li><a href="#{{child.href|safe}}">{{child.name}}</a></li>
+                                            {% endfor %}
+                                        </ul> {# #}
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/templates/sidebar.html:32]: Can be:

```suggestion
                                        </ul>
```

File: src/librustdoc/html/templates/sidebar.html [hunk 3]
```diff
@@ -11,24 +6,44 @@ <h2 class="location"> {# #}
...
+                                    {% endif %}
+                                </li> {# #}
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/templates/sidebar.html:34]: Can be:

```suggestion
                                </li>
```

File: src/librustdoc/html/templates/sidebar.html [hunk 4]
```diff
@@ -11,24 +6,44 @@ <h2 class="location"> {# #}
...
                 {% endif %}
             {% endfor %}
         </section>
     {% endif %}
+    <div id="ModNav">
     {% if !path.is_empty() %}
-        <h2><a href="{% if is_mod %}../{% endif %}index.html">In {{+ path}}</a></h2>
+        <h2{% if parent_is_crate +%} class="in-crate"{% endif %}> {# #}
+            <a href="{% if is_mod %}../{% endif %}index.html">In {{+ path}}</a> {# #}
+        </h2> {# #}
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/templates/sidebar.html:46]: Same here.

**Comment by GuillaumeGomez** [comment]
Just a few small things:

* It's hard to spot the separation between the ToC and the module items. Could you add a title "Module items" maybe?
* backticks are kept in the table (can be seen for [this heading](http://notriddle.com/rustdoc-html-demo-9/toc/rust-derive-builder/derive_builder/index.html#no_std-support-on-nightly)). I don't think it should be there (or any markdown syntax).
* I spotted some nits but they can be ignored, it's really detail and I'm not convinced it's really any better.
* Can you add an HTML test in case the section is empty (to ensure the HTML element is not generated) please?

Apart from that, looks good to me! Once done, I'll approve.

File: tests/rustdoc/sidebar/top-toc-html.rs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+// ignore-tidy-linelength
+
+#![crate_name = "foo"]
+#![feature(lazy_type_alias)]
+#![allow(incomplete_features)]
+
+//! # Basic [link](https://example.com) and *emphasis* and `code`
```
Comment (GuillaumeGomez) [inline tests/rustdoc/sidebar/top-toc-html.rs:7]: Can you an embed case as well please? Like `**_very rich_**`.

**Comment by GuillaumeGomez** [comment]
When I enable "Hide table of contents", it also hides "Crate items". Is it expected? If yes, could it just hide the table of contents instead?

File: tests/rustdoc/sidebar/module.rs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+#![crate_name = "foo"]
+
+//@ has 'foo/index.html'
+//@ has - '//section[@id="TOC"]/h3' 'Crate Items'
```
Comment (GuillaumeGomez) [inline tests/rustdoc/sidebar/module.rs:4]: You added the test for the two headings, awesome!

File: src/librustdoc/html/markdown.rs [hunk 7]
```diff
@@ -1413,7 +1429,11 @@ impl MarkdownWithToc<'_> {
             html::push_html(&mut s, p);
         }
 
-        format!("<nav id=\"TOC\">{toc}</nav>{s}", toc = toc.into_toc().print())
+        (toc.into_toc(), s)
+    }
+    pub(crate) fn into_string(self) -> String {
+        let (toc, s) = self.into_parts();
+        format!("<nav id=\"TOC\">{toc}</nav>{s}", toc = toc.print())
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs:1436]: Just one thought: this ID and `ModNav` are the only IDs with capital letters. A reason for that?

Comment (GuillaumeGomez) [reply src/librustdoc/html/markdown.rs:1436]: Yes please. OCD triggering hard on this one. ^^'

Comment (GuillaumeGomez) [reply src/librustdoc/html/markdown.rs:1436]: Oh I see. Can you update `ModNav` at least please?

Comment (GuillaumeGomez) [reply src/librustdoc/html/markdown.rs]: 👍

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me now. Thanks for this feature!

@rfcbot reviewed

**Comment by GuillaumeGomez** [comment]
Thanks everyone!

@bors r=t-rustdoc

**Pull request merged**